### PR TITLE
[MM-27132] Get user group mention keys for channel

### PR DESCRIPTION
**Pull request by fmunshi**
#### Summary
- Modify user mention keys selector to look at group mention keys associated to that specific channel
- Move group mentions permission checks out of redux as certain selectors require the user to get the list of groups even if they cannot mention (eg. highlighting)

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-27132

#### Related Pull Requests
- https://github.com/mattermost/mattermost-redux/pull/1214

### CODE REVIEW

File: components/create_post/index.js [hunk 1]
```diff
@@ -93,6 +91,7 @@ function makeMapStateToProps() {
         });
         const channelMemberCountsByGroup = selectChannelMemberCountsByGroup(state, currentChannel.id);
         const currentTeamId = getCurrentTeamId(state);
+        const groupsWithAllowReference = useGroupMentions ? new Map() : getAssociatedGroupsForReferenceByMention(state, currentTeamId, currentChannel.id);
```
Comment (catalintomai) [inline components/create_post/index.js:94]: Farhan, can you help me understand this? If we do not have `groupMentions` set, then we call `getAssociatedGroupsForReferenceByMention`?

Comment (fmunshi) [reply components/create_post/index.js]: This was incorrect... good catch @catalintomai 😅

File: components/create_comment/index.js [hunk 2]
```diff
@@ -76,6 +74,7 @@ function makeMapStateToProps() {
             permission: Permissions.USE_GROUP_MENTIONS,
         });
         const channelMemberCountsByGroup = selectChannelMemberCountsByGroup(state, ownProps.channelId);
+        const groupsWithAllowReference = useGroupMentions ? getAssociatedGroupsForReferenceByMention(state, channel.team_id, channel.id) : new Map();
```
Comment (hmhealey) [inline components/create_comment/index.js:77]: We can't pass in a new object when we're not using group mentions here or else CreateComment will constantly re-render unnecessarily. It did that before with the previous code as well, but we should fix it while we're in here. Instead, we should pass in `null` and handle it in the component.

File: components/create_post/index.js [hunk 3]
```diff
@@ -93,6 +91,7 @@ function makeMapStateToProps() {
         });
         const channelMemberCountsByGroup = selectChannelMemberCountsByGroup(state, currentChannel.id);
         const currentTeamId = getCurrentTeamId(state);
+        const groupsWithAllowReference = useGroupMentions ? getAssociatedGroupsForReferenceByMention(state, currentTeamId, currentChannel.id) : new Map();
```
Comment (hmhealey) [inline components/create_post/index.js:94]: Same here

File: components/textbox/index.ts [hunk 4]
```diff
@@ -30,12 +32,20 @@ const makeMapStateToProps = () => {
 
     return (state: GlobalState, ownProps: Props) => {
         const teamId = getCurrentTeamId(state);
+        const license = getLicense(state);
+        const useGroupMentions = license?.IsLicensed === 'true' && license?.LDAPGroups === 'true' && haveIChannelPermission(state, {
+            channel: ownProps.channelId,
+            team: teamId,
+            permission: Permissions.USE_GROUP_MENTIONS,
+        });
+        const autocompleteGroups = useGroupMentions ? getAssociatedGroupsForReference(state, teamId, ownProps.channelId) : [];
```
Comment (hmhealey) [inline components/textbox/index.ts:41]: Same here. You can also make a selector to return a memoized copy of [] instead

File: components/create_comment/index.js (hunk 2 above)
Comment (mkraft) [reply components/create_comment/index.js]: Updated. Thanks.

File: components/create_post/index.js (hunk 3 above)
Comment (mkraft) [reply components/create_post/index.js]: Updated.

File: components/textbox/index.ts (hunk 4 above)
Comment (mkraft) [reply components/textbox/index.ts]: Updated.

**Comment by mkraft** [comment]
Taking over this PR for @fmunshi while he's on PTO.

**Comment by fmunshi** [comment]
@srkgupta I have fixed the issue mentioned above.

**Comment by amyblais** [comment]
Rohitesh is off this week so I've swapped the QA reviewer to Ogi,

**Comment by fmunshi** [comment]
Hi @ogi-m, In order to test this you will need a user that belongs to two different ldap groups (both with allow_reference set to true), and a group synced channel that only has _one_ of those two groups synced to it. 

Once you have that the expected scenario can be reproduced, when the synced group is mentioned it should be highlighted and when the non synced group is @ mentioned it should just appear as a blue keyword and not have any highlight

**Review by ogi-m — APPROVED** [review]
Tested and passed - outside group is highlighted in blue 👍

**Pull request merged**